logs/querysample/querysample.go

```diff
@@ -116,3 +133,19 @@ func TransformLogMinDurationStatementToQuerySample(logLine state.LogLine, queryT
 	}
 	return sample, true
 }
+
+func findQueryParameters(paramText string) []null.String {
+	// Handle Query Parameters (available from Postgres 16+)
+	var parameters []null.String
+	// Regular expression to find all values in single quotes or NULL
+	// Query Parameters example: $1 = 'foo', $2 = '123', $3 = NULL, $4 = 'bo''o'
+	re := regexp.MustCompile(`'((?:[^']|'')*)'|NULL`)
+	for _, part := range re.FindAllString(paramText, -1) {
+		if part == "NULL" {
+			parameters = append(parameters, null.NewString("", false))
+		} else {
+			parameters = append(parameters, null.StringFrom(strings.Trim(part, "'")))
```

Comment:
Does this mean that for parameters with escaped quotes (e.g., `o''clock`), we store them with those escapes (i.e., the string that ends up in the snapshot is `o''clock` rather than the actual parameter value `o'clock`? Is that what we are doing for `log_min_duration_statement` parameters? I know that the code here is similar to the code you linked above, but I tried to trace back where the parameter values come from there, checking to confirm how escaped quotes are handled there, and I got lost. Are we consistent?

Reply:
Still wondering about this.

Reply:
You don't realize that you have 2 weeks old pending comment until someone else wonders that you haven't said anything 🙃

Reply:
Interestingly, it does look like I commented before Maciek commented, but for the record, Maciek did wonder before I commented.

Reply:
Oh whoops, I'm not sure how I missed that. I think GH did not refresh the comments when I was doing my review somehow.

> Regarding storing param data as is (o''clock) or not (o'clock), I prefer to keep it as is. Honestly, nothing is going to be perfect around here (e.g. we don't have types, some binary params will act weirdly, etc.), so I'd like to start from somewhere. If we found some inconvenience, we can always come back here to improve things.

Okay, but I do think we need to be really careful here. This is the kind of thing that leads to SQL injection, and while that's not a problem in this situation, I think giving the impression that we don't know how to handle SQL injection could make our customers lose trust.

Reply:
I see, that's a good point. Basically, we'll need to do the opposite of this here to get the actual value:
https://github.com/postgres/postgres/blob/b4db64270e0c18f72c9c7bf214c809949875a85a/src/backend/nodes/params.c#L321-L336

Though, given that we are only going to pass this value as a string to pganalyze side, how much we're going to mutate this value in the collector side would be question. IOW, if we pass this as is, we could treat these parameters as "param log string" (not a param) and do all the potential parsing in pganalyze app side. With that, I think we can say that we're not neglecting the SQL injection aspect.

-----

Regarding "unescape", here is some example of complex parameter value:

```
$4 = '{\"(12345,-56789,111111,\\\"\\\\\\\\x041d44a78f95610f\\\",c15c4fd4-f643-42db-ac03-9335a6bd78fc)\"}'
```

It's hard to read, but it's one element array of record. The record is with the first 3 values being integer, 4th value is bytea, 5th value is uuid. This is an extreme example (note: this is an argument of function, FYI), but how should we treat the case like this? I think we need to do some dirty thing case by case, and if we want to reuse this as a parameter to be used for the EXPLAIN command, more manipulations are needed, which I feel like good to have it in pganalyze side, not in the collector side.

Reply:
Great point. I think doing any filtering on the receiving side is a much better idea. That way, any updates or fixes to the parameter filtering logic can be deployed immediately, without having to wait for users to upgrade their collectors.